README.md
Generate and save QR Code Image.

<img src="https://i.ibb.co/8xg3fkg/demogif.gif" alt="Package usage demo"/>

---

## Basic Usage

```dart

    final generator = QRGenerator();

    await generator.generate(
      data: 'Hello World!',
      filePath: '$outputDir/hello.png',
    );

```

## Full Usage

```dart

  Future saveQRImage() async {

    final generator = QRGenerator();

    await generator.generate(
      data: 'Hello World!',
      filePath: '$outputDir/hello.png',
      scale: 10,
      foregroundColor: Colors.yellow,
      backgroundColor: Colors.blue,
      errorCorrectionLevel: ErrorCorrectionLevel.medium,
      qrVersion: 4,
    );
  }

```

The Image will get saved to the provided `filePath`.

---

### TODO

- [ ] Add Transperency Support
- [ ] Write Documentation
- [ ] Add Image in Qr Code
- [ ] Add Tests


CHANGELOG.md
## 0.2.9

- **FIX** Added missing await to async function ([#1](https://github.com/Manty-K/QR-Image-Generator/commit/290874a91590bd573d6dbd024db9cdb56c3cb0bd))

  Thanks to @MesserschmittX

## 0.2.8

- Bug Fix

## 0.2.7

- Optimised QR Generation Algorithm
- `path_provider` dependency removed

## 0.2.6

- Analyzer issue fix

## 0.2.5

- Readme update

## 0.2.4

- Unknown Platform bug fix

## 0.2.3

- Removed Flutter dependancy

## 0.2.2

- Error Correction Level control added
- Set QR Version from data

## 0.2.1

- Background and foreground color support addded.
- Padding Feature added.
- Documentation added.

## 0.2.0

- Flutter Example Added.

## 0.1.0

- Initial version.


test/q_gene_test.dart
import 'package:qr_image_generator/qr_image_generator.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    setUp(() {
      // Additional setup goes here.
    });

    test('First Test', () {});
  });
}


example/lib/main.dart
import 'package:file_picker/file_picker.dart';
import 'package:flutter/material.dart';
import 'package:qr_image_generator/qr_image_generator.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'QR Image Generator Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'QR Image Generator Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TextEditingController textEditingController = TextEditingController();
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Column(
            mainAxisSize: MainAxisSize.min,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text('Enter Text',
                  style: Theme.of(context).textTheme.displaySmall),
              const SizedBox(height: 10),
              TextFormField(
                controller: textEditingController,
                textAlign: TextAlign.center,
                decoration: const InputDecoration(
                  border: OutlineInputBorder(),
                ),
              ),
              const SizedBox(height: 100),
              ElevatedButton(
                onPressed: saveQRImage,
                child: const Text('Save QR'),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Future saveQRImage() async {
    FocusScope.of(context).unfocus();
    String? filePath = await FilePicker.platform.saveFile(
      fileName: 'demoQr.png',
      type: FileType.image,
    );
    if (filePath == null) {
      return;
    }

    final generator = QRGenerator();

    await generator.generate(
      data: textEditingController.text,
      filePath: filePath,
      scale: 10,
      padding: 0,
      foregroundColor: Colors.yellow,
      backgroundColor: Colors.blue,
      errorCorrectionLevel: ErrorCorrectionLevel.medium,
    );
  }
}


lib/qr_image_generator.dart
library qr_image_generator;

import 'dart:io';
import 'dart:ui';
import 'package:path_provider/path_provider.dart';
import 'package:qr/qr.dart';
import 'package:image/image.dart' as img;

class QRGenerator {
  late String _selectedData;

  late List<List<bool>> _imageData;

  late String _outputFilePath;

  late int _scale;

  late int _padding;

  late Color _bgColor;
  late Color _fgColor;

  late ErrorCorrectionLevel _errorCorrectionLevel;

  String? _tempDirPath;

  int? _qrVersion;

  String get _tempFilePath {
    final splitted = _outputFilePath.split('/');

    final filename = splitted.removeLast();

    final newFilename = 'temp_$filename';

    final filepath = '$_tempDirPath/$newFilename';

    return filepath;
  }

  /// Generate and save QR Code
  ///
  /// String [data] to be converted to QR Code.
  ///
  /// The image will be saved at [filePath].
  ///
  /// [padding] is size of a QR Module. Default is 1.
  ///
  /// [backgroundColor] color of a QR. Defaults to [Colors.white].
  ///
  /// [foregroundColor] of a QR. Defaults to [Colors.black].
  ///
  /// Choose [errorCorrectionLevel] between low, medium, quartile and high. Defaults to [ErrorCorrectionLevel.medium].
  Future<String> generate({
    required String data,
    required String filePath,
    int scale = 5,
    int padding = 1,
    Color backgroundColor = const Color.fromRGBO(255, 255, 255, 1),
    Color foregroundColor = const Color.fromRGBO(0, 0, 0, 1),
    ErrorCorrectionLevel errorCorrectionLevel = ErrorCorrectionLevel.medium,

    /// Generator automatically sets version from data. But you can also explicitly set qr version [1-40];
    int? qrVersion,
  }) async {
    // ? Use assert statements

    if (padding <= 0) {
      throw 'Padding should be more than 0';
    }

    if (scale <= 0) {
      throw 'Scale should be more than 0';
    }

    if (data.trim().isEmpty) {
      throw 'Data should not be empty';
    }

    if (!filePath.endsWith('.png')) {
      throw 'File should end with .png';
    }

    if (qrVersion != null) {
      if (qrVersion > 40 || qrVersion < 1) {
        throw 'QR version should be 1 - 40';
      }
    }

    _selectedData = data;
    _outputFilePath = filePath;
    _scale = scale;
    _padding = padding;
    _bgColor = backgroundColor;
    _fgColor = foregroundColor;
    _errorCorrectionLevel = errorCorrectionLevel;
    _qrVersion = qrVersion;

    if (_tempDirPath == null) {
      final tempDir = await getTemporaryDirectory();
      _tempDirPath = tempDir.path;
    }

    try {
      await _makeImage();
      return filePath;
    } catch (e) {
      throw 'Generate Image Error';
    }
  }

  Future<void> _makeImage() async {
    late QrCode qr;

    if (_qrVersion == null) {
      qr = QrCode.fromData(
          data: _selectedData,
          errorCorrectLevel: _errorCorrectionLevelInt(_errorCorrectionLevel));
    } else {
      qr = QrCode(_qrVersion!, _errorCorrectionLevelInt(_errorCorrectionLevel))
        ..addData(_selectedData);
    }

    final qrImage = QrImage(qr);

    _imageData = _getModules(qrImage);

    await _makeMiniImage();

    await _enlarge();
  }

  Future<void> _makeMiniImage() async {
    final data = _imageData;

    final spacing = _padding;

    final image = img.Image(
      width: data.length + (spacing * 2),
      height: data.length + (spacing * 2),
    );

    ///Background mapping
    for (int i = 0; i < data.length + (spacing * 2); i++) {
      for (int j = 0; j < data.length + (spacing * 2); j++) {
        image.setPixel(i, j, _convertMaterialColorToImageColor(_bgColor));
      }
    }

    /// QR Code Mapping
    for (int i = 0; i < data.length; i++) {
      final row = data[i];
      for (int j = 0; j < row.length; j++) {
        final d = row[j];

        if (d) {
          /// Foreground Color
          img.Color c = _convertMaterialColorToImageColor(_fgColor);
          image.setPixel(i + spacing, j + spacing, c);
        }
      }
    }

    final png = img.encodePng(image);
    await File(_tempFilePath).writeAsBytes(png);
  }

  Future _enlarge() async {
    final imagePath = _tempFilePath;
    final cmd = img.Command()
      ..decodeImageFile(imagePath)
      ..copyResize(width: _imageData.length * _scale)
      ..writeToFile(_outputFilePath);

    await cmd.executeThread();

    _deleteMini();
  }

  Future<void> _deleteMini() async {
    await File(_tempFilePath).delete();
  }

  List<List<bool>> _getModules(QrImage image) {
    List<List<bool>> qrdata = [];

    /// Add empty lists
    for (int i = 0; i < image.moduleCount; i++) {
      qrdata.add([]);
    }

    /// Add Data
    for (var x = 0; x < image.moduleCount; x++) {
      for (var y = 0; y < image.moduleCount; y++) {
        if (image.isDark(y, x)) {
          // render a dark square on the canvas
          qrdata[x].add(true);
        } else {
          qrdata[x].add(false);
        }
      }
    }
    return qrdata;
  }
}

img.Color _convertMaterialColorToImageColor(Color c) {
  return img.ColorRgba8(c.red, c.green, c.blue, c.alpha);
}

int _errorCorrectionLevelInt(ErrorCorrectionLevel level) {
  switch (level) {
    case ErrorCorrectionLevel.low:
      return QrErrorCorrectLevel.L;
    case ErrorCorrectionLevel.medium:
      return QrErrorCorrectLevel.M;
    case ErrorCorrectionLevel.quartile:
      return QrErrorCorrectLevel.Q;
    case ErrorCorrectionLevel.high:
      return QrErrorCorrectLevel.H;
  }
}

/// QR Code has error correction capability to restore data if the code is dirty or damaged.
/// Four error correction levels are available for users to choose according to the operating environment.
/// Raising this level improves error correction capability but also increases the amount of data QR Code size.
/// To select error correction level, various factors such as the operating environment and QR Code size need to be considered.
/// [ErrorCorrectionLevel.quartile] or [ErrorCorrectionLevel.high] may be selected for factory environment where QR Code get dirty,
/// whereas [ErrorCorrectionLevel.low] may be selected for clean environment with the large amount of data.
/// Typically, [ErrorCorrectionLevel.medium] (15%) is most frequently selected.
enum ErrorCorrectionLevel {
  low,
  medium,
  quartile,
  high,
}


lib/src/qr_image_generator_base.dart
import 'dart:io';
import 'dart:ui';

import 'package:image/image.dart' as img;
import 'package:qr/qr.dart';

class QRGenerator {
  late String _selectedData;

  late List<List<bool>> _imageData;

  late String _outputFilePath;

  late int _scale;

  late int _padding;

  late Color _bgColor;
  late Color _fgColor;

  late ErrorCorrectionLevel _errorCorrectionLevel;

  int? _qrVersion;

  /// Generate and save QR Code
  Future<String> generate({
    /// String [data] to be converted to QR Code.
    required String data,

    /// The image will be saved at [filePath].
    required String filePath,

    /// Scale factor of QR Code. Default is 5.
    int scale = 5,

    /// [padding] factor around QR Code. Default is 1.
    int padding = 1,

    /// [backgroundColor] color of a QR. Defaults to [Colors.white].
    Color backgroundColor = const Color.fromRGBO(255, 255, 255, 1),

    /// [foregroundColor] of a QR. Defaults to [Colors.black].
    Color foregroundColor = const Color.fromRGBO(0, 0, 0, 1),

    /// Choose [errorCorrectionLevel] between low, medium, quartile and high. Defaults to [ErrorCorrectionLevel.medium].
    ErrorCorrectionLevel errorCorrectionLevel = ErrorCorrectionLevel.medium,

    /// Generator automatically sets version from data. But you can also explicitly set qr version [1 - 40];
    int? qrVersion,
  }) async {
    // ? Use assert statements

    if (padding < 0) {
      throw 'Padding should not be negative';
    }

    if (scale <= 0) {
      throw 'Scale should be more than 0';
    }

    if (data.trim().isEmpty) {
      throw 'Data should not be empty';
    }

    if (!filePath.endsWith('.png')) {
      throw 'Filename should end with .png';
    }

    if (qrVersion != null) {
      if (qrVersion > 40 || qrVersion < 1) {
        throw 'QR version should be 1 - 40';
      }
    }

    _selectedData = data;
    _outputFilePath = filePath;
    _scale = scale;
    _padding = padding;
    _bgColor = backgroundColor;
    _fgColor = foregroundColor;
    _errorCorrectionLevel = errorCorrectionLevel;
    _qrVersion = qrVersion;

    try {
      await _makeImage();
      return filePath;
    } catch (e) {
      throw 'Generate Image Error';
    }
  }

  Future<void> _makeImage() async {
    late QrCode qr;

    if (_qrVersion == null) {
      qr = QrCode.fromData(
          data: _selectedData,
          errorCorrectLevel: _errorCorrectionLevelInt(_errorCorrectionLevel));
    } else {
      qr = QrCode(_qrVersion!, _errorCorrectionLevelInt(_errorCorrectionLevel))
        ..addData(_selectedData);
    }

    final qrImage = QrImage(qr);

    _imageData = _getModules(qrImage);

    await _make();
  }

  Future _make() async {
    /// Define Image
    final size = (_imageData.length * _scale) + (_padding * _scale * 2);

    final image = img.Image(
      width: size,
      height: size,
    );

    // added padding
    final imageDataWithPadding = _addPadding();

    /// scale data
    final enlargedData = _MarticScaler<bool>()
        .scale(data: imageDataWithPadding, scaleFactor: _scale);

    for (int i = 0; i < enlargedData.length; i++) {
      final row = enlargedData[i];

      for (int j = 0; j < enlargedData.length; j++) {
        final d = row[j];

        if (d) {
          img.Color c = _convertMaterialColorToImageColor(_fgColor);

          image.setPixel(i, j, c);
        } else {
          image.setPixel(i, j, _convertMaterialColorToImageColor(_bgColor));
        }
      }
    }
    final png = img.encodePng(image);
    await File(_outputFilePath).writeAsBytes(png);
  }

  List<List<bool>> _addPadding() {
    List<List<bool>> out = [];

    for (int i = 0; i < _padding; i++) {
      final l = List.generate(_imageData.length + (_padding * 2), (_) => false);
      out.add(l);
    }

    for (final x in _imageData) {
      List<bool> myList = x;
      for (int i = 0; i < _padding; i++) {
        myList = [false, ...myList, false];
      }

      out.add(myList);
    }

    for (int i = 0; i < _padding; i++) {
      final l = List.generate(_imageData.length + (_padding * 2), (_) => false);
      out.add(l);
    }

    return out;
  }

  List<List<bool>> _getModules(QrImage image) {
    List<List<bool>> qrdata = [];

    /// Add empty lists
    for (int i = 0; i < image.moduleCount; i++) {
      qrdata.add([]);
    }

    /// Add Data
    for (var x = 0; x < image.moduleCount; x++) {
      for (var y = 0; y < image.moduleCount; y++) {
        if (image.isDark(y, x)) {
          // render a dark square on the canvas
          qrdata[x].add(true);
        } else {
          qrdata[x].add(false);
        }
      }
    }

    return qrdata;
  }
}

img.Color _convertMaterialColorToImageColor(Color c) {
  return img.ColorRgba8(c.red, c.green, c.blue, c.alpha);
}

int _errorCorrectionLevelInt(ErrorCorrectionLevel level) {
  switch (level) {
    case ErrorCorrectionLevel.low:
      return QrErrorCorrectLevel.L;
    case ErrorCorrectionLevel.medium:
      return QrErrorCorrectLevel.M;
    case ErrorCorrectionLevel.quartile:
      return QrErrorCorrectLevel.Q;
    case ErrorCorrectionLevel.high:
      return QrErrorCorrectLevel.H;
  }
}

/// QR Code has error correction capability to restore data if the code is dirty or damaged.
/// Four error correction levels are available for users to choose according to the operating environment.
/// Raising this level improves error correction capability but also increases the amount of data QR Code size.
/// To select error correction level, various factors such as the operating environment and QR Code size need to be considered.
/// [ErrorCorrectionLevel.quartile] or [ErrorCorrectionLevel.high] may be selected for factory environment where QR Code get dirty,
/// whereas [ErrorCorrectionLevel.low] may be selected for clean environment with the large amount of data.
/// Typically, [ErrorCorrectionLevel.medium] (15%) is most frequently selected.
enum ErrorCorrectionLevel {
  low,
  medium,
  quartile,
  high,
}

class _MarticScaler<T> {
  List<List<T>> scale({required List<List<T>> data, required int scaleFactor}) {
    _setInitialData(data, scaleFactor);
    _iterateOverRows();

    return output;
  }

  void _setInitialData(List<List<T>> data, int scale) {
    _initialData = data;
    _moduleCount = data.length;
    _scale = scale;
  }

  late List<List<T>> _initialData;

  late int _scale;

  late int _moduleCount;

  List<List<T>> output = [];

  void _iterateOverRows() {
    for (int i = 0; i < _moduleCount; i++) {
      final lis = _initialData[i];

      _fillRows(lis);
    }
  }

  void _fillRows(List list) {
    final List<List<T>> elementScaled = [];

    // Fill empty
    for (int s = 0; s < _scale; s++) {
      elementScaled.add([]);
    }

    // Fill row
    for (int i = 0; i < _scale; i++) {
      for (int j = 0; j < list.length; j++) {
        final k = List.generate(_scale, (_) => list[j]);
        for (final e in k) {
          elementScaled[i].add(e);
        }
      }
    }

    // Add row to output
    for (final element in elementScaled) {
      output.add(element);
    }
  }
}
